contracts/src/gateway/router/CheckpointingFacet.sol

```diff
@@ -52,9 +52,42 @@ contract CheckpointingFacet is GatewayActorModifiers {
         bytes32 membershipRootHash,
         uint256 membershipWeight
     ) external systemActorOnly {
-        if (checkpoint.blockHeight % s.bottomUpCheckPeriod != 0) {
-            revert InvalidCheckpointEpoch();
-        }
+        // TODO: Discussion:
+        //
+        // This check is removed:
+        //
+        // if (checkpoint.blockHeight % s.bottomUpCheckPeriod != 0) {
+        //    revert InvalidCheckpointEpoch();
+        // }
+        //
+        // First of all, the check is buggy in the first place as it does not support
+        // a full message batch before a checkpoint period is reached.
+        // Then, `prevBottomUpCheckpointHeight` is introduced, as long as BU checkpoints
+        // can form a chain, it should be fine.
+        //
+        // Instead, one can add the following check:
+        //
+        // if (checkpoint.prevHeight != s.prevBottomUpCheckpointHeight) {
```

Comment:
```suggestion
        // if (checkpoint.prevHeight != s.lastBottomUpCheckpointHeight) {
```
I remember finding our wording confusing somewhere, I think it was _current_ vs _last_ validator set. 

Anyway, I think the `s` means the current state, and in that context it's "the last bottom-up checkpoint we received". If we use `prev` it's a bit less clear to me if there is something newer.